### ISSUE TITLE
Clarify folded `try` text format

### ISSUE THREAD

**Issue by binji** [WebAssembly/exception-handling#52]
The proposal describes the flat format for a try block, but not the folded format. I believe the previous proposal used:

```
(try
  ...
  (catch
    ...
  )
)
```

Should we do the same here?

**Comment by eholk** [comment]
For the s-expression format, I like what you've listed (although normally closing parentheses go on the same line). In the proposal and a lot of the discussions we've been using a more Pythonesque format because that seems to be easier for most people to read.

**Comment by binji** [comment]
Sorry, a little context: I was starting to work on parsing this in wabt, so we need to support some sort of folded text format. Most blocks only have one section, so they just look like: `(block ...)`. If has two sections, and has sugar to simplify single expression true/false branches as well:

```
(if (cond) (nop) (nop))
(if (cond) (then ...) (else ...))

;; you can even do bizarre stuff like
(cond)
(if (then ...) (else ...))
```

Try/catch could work this way, but then you'd need another keyword, I suppose:

```
(try (do ...) (catch ...))
```

I don't think this is a good idea, but I thought I'd mention it.

**Comment by eholk** [comment]
Would just `(try ... (catch ...))` work?

**Comment by binji** [comment]
Yes, that's what I've currently implemented and I didn't notice any issues.

**Comment by aheejin** [comment]
@binji Do we support syntactic sugar `then`? (I didn't know) Is it official? Is it converted to a nop later?

@rossberg In your suggestion, `do` is gonna be converted to a nop as well?

**Comment by rossberg** [comment]
@aheejin, yes the `then` syntax is official. It is just part of the structured textual syntax of the `if` instruction, it does not correspond to a specific byte in the binary format (in the same way a keyword like, say, `param` doesn't).

**Comment by aheejin** [comment]
@rossberg Then can we assume we have `then` for `if_except` too?

**Comment by aheejin** [comment]
For the folded `try` text format, I had assumed `(try ... (catch ...))` before. But I just started to wonder, if a `catch` block is enclosed in a `try`, this seems to imply the value stack from the `try` block should not be popped when a `catch` block is entered. But the current proposal says

> Once a catching try block is found for the thrown exception, the operand stack is popped back to the size the operand stack had when the try block was entered, and then an except_ref referring to the caught exception is pushed back onto the operand stack.

which sounds like `catch` and `try` are separate blocks rather than `catch` is enclosed within `try`.

**Comment by rossberg** [comment]
@aheejin, both blocks are part of the `try`, but they are separate indeed, neither is nested in the other. What you describe is what I was trying to get at in my earlier comment.

**Comment by aheejin** [comment]
By the way, if we are going to introduce one more parenthesis for `catch`, would that affect the argument of branches within `catch` block? For example,

```wast
(try
 ...
 (catch
  ...
  br N
 )
)
some code
```

To get to the `some code` part, what `N` should we use? 0? 1?

**Comment by rossberg** [comment]
@aheejin, no, don't worry, it's just concrete syntax, no bearing on semantics (cf. `if`, where this isn't the case either).

**Comment by aheejin** [comment]
Currently both binaryen and wabt implement the folded `try` text format OP suggested:
```
(try
  ...
  (catch
    ...
  )
)
```

So closing this for now. Please feel free to reopen if necessary.

**Comment by aheejin** [comment]
In https://github.com/WebAssembly/binaryen/pull/2780#discussion_r419688507, @binji suggested to reopen this issue and we make `do` mandatory, as @rossberg suggested above, so the text format is
```
(try (do ...) (catch ...))
```
I'm not against it. What do people think?

**Comment by rossberg** [comment]
It would certainly improve syntactic consistency. :)

Just to clarify, the folded form would be `(try (do ...) (catch ...))` and the flat one `try ... catch ... end`, analogous to `if`.

**Comment by aheejin** [comment]
If no one has concerns, I'll add `do` in the `try` forded text format soon. Let me know if you have any.

**Comment by aheejin** [comment]
One more thing to clarify: Should we allow an empty `try` body? For example, when there's no instructions in the `try` body, should we do
```wast
(try
  (catch
    ...
  )
)
```
or
```wast
(try
  (do)
  (catch
    ...
  )
)
```

Currently Binaryen naturally allows an empty `try` body because we don't have `do` anyway, but we have to think about it after adding `do`. I'm still inclined to allow an empty body of `try` (so no empty `do`) when there's no instructions in a `try` body.

**Comment by rossberg** [comment]
For `if`, we don't allow to omit `then`, though we allow to omit `else` (as in the binary format).

For `try`, an empty body is just useless (makes the `try` equivalent to `nop`), so there seems little reason to introduce a specific syntactic shorthand for that.

**Comment by aheejin** [comment]
So you mean we can allow `try` body to be empty (without an empty `(do)`), right?

**Comment by rossberg** [comment]
No, the opposite, since that would be introducing a (useless) extra grammar case. ;)

**Comment by aheejin** [comment]
Implemented in WebAssembly/Binaryen#2846, WebAssembly/Binaryen#2851, and WebAssembly/wabt#1425. Closing.
